### PR TITLE
Reference resolution of custom object `key-value-document` references (and set of) as an attribute of nested type.

### PR DESCRIPTION
**Pull request by ahmetoz**
#### Summary
#564 

#### Todo

- [x] Validations.
- Tests
    - [X] Unit  
    - [X] Integration
- [ ] Documentation
- [X] Add Release Notes entry.
- [x] Coverage.

### CODE REVIEW

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithNestedReferencedCustomObjectsIT.java [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+
+            assertThat(nestedAttributeNameField.asText()).isEqualTo(CUSTOM_OBJECT_REFERENCE_ATTR_NAME);
+            assertThat(nestedAttributeValueField.get(REFERENCE_TYPE_ID_FIELD).asText())
+                .isEqualTo(CustomObject.referenceTypeId());
+            assertThat(nestedAttributeValueField.get(REFERENCE_ID_FIELD).asText()).isEqualTo(testCustomObject2.getId());
+        });
+    }
+
+    @Test
+    void sync_withNonExistingNestedCustomObjectReferenceAsAttribute_ShouldFailCreatingTheProduct() {
```
Comment (JudeNiroshan) [inline src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithNestedReferencedCustomObjectsIT.java:420]: So it means, we expect in _project-sync_, we execute `CustomObjectSyncer` before we execute `ProductSyncer`?

File: src/test/java/com/commercetools/sync/products/helpers/ProductBatchValidatorTest.java [hunk 2]
```diff
@@ -376,6 +380,43 @@ void validateAndCollectReferencedKeys_WithDrafts_ShouldValidateCorrectly() {
             format(PRODUCT_VARIANT_DRAFT_SKU_NOT_SET, 1, inValidProductDraft2.getKey()));
     }
 
+    @Test
+    void validateAndCollectReferencedKeys_WithKeyValueDocumentAttDrafts_ShouldHaveKeyValidationError() {
```
Comment (JudeNiroshan) [inline src/test/java/com/commercetools/sync/products/helpers/ProductBatchValidatorTest.java:384]:
```suggestion
    void validateAndCollectReferencedKeys_WithKeyValueDocumentAttDrafts_ShouldNotHaveKeyValidationError() {
```

**Review by JudeNiroshan — APPROVED** [review]

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithNestedReferencedCustomObjectsIT.java (hunk 1 above)
Comment (ahmetoz) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithNestedReferencedCustomObjectsIT.java:420]: yes, I haven't' check the custom object part on the project sync, I assume @lojzatran is considering it.

Comment (lojzatran) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/products/ProductSyncWithNestedReferencedCustomObjectsIT.java:420]: I do it like that in the project-sync (custom object before product), but this actually should not matter as they're not referenced to each other.

**Pull request merged**